### PR TITLE
ARROW-8421: [Rust] [Parquet] Merge in-progress writer & reader

### PR DESCRIPTION
**Pull request by nevi-me**
This merges the in-progress Arrow <> Parquet IO branch into master, so we can continue with development here.

Please do not squash, as we'd like to preserve the individual commits as they're already squashed into their respective JIRAs.

### CODE REVIEW

**Comment by jorgecarleitao** [comment]
There was probably some mis-understanding during the call yesterday, but I was expecting this to be a pull request and approved by another committer _before being merged_. This merge IMO did not follow our review practice and did not gave time to other committers to go through it.

File: rust/arrow/src/array/data.rs [hunk 1]
```diff
@@ -209,6 +209,61 @@ impl ArrayData {
...
+    right_offset: usize,
+) {
+    // for convenience, we assume that the buffer lengths are only unequal if one has padding,
+    // so we take the shorter length so we can discard the padding from the longer length
+    let shorter_len = left.len().min(right.len());
+    let s_sliced = left.bit_slice(left_offset, shorter_len);
+    let o_sliced = right.bit_slice(right_offset, shorter_len);
+    assert_eq!(s_sliced, o_sliced);
+}
+
```
Comment (jorgecarleitao) [inline rust/arrow/src/array/data.rs:266]: This is not correct, see [this comment](https://github.com/apache/arrow/pull/8546#discussion_r513933731)

Comment (jorgecarleitao) [inline rust/arrow/src/array/data.rs:266]: IMO this is not correct, see [this comment](https://github.com/apache/arrow/pull/8546#discussion_r513933731)

**Review by jorgecarleitao — COMMENTED** [review]
I am a bit lost here: I though that, we would still conduct a review of this code before a merge, as no-one was reviewing while the branch was up.

IMO we introduced errors to the `arrow` crate, and there was no opportunity for others to review before this merge.

**Comment by nevi-me** [comment]
> There was probably some mis-understanding during the call yesterday, but I was expecting this to be a pull request and approved by another committer _before being merged_. This merge IMO did not follow our review practice and did not gave time to other committers to go through it.

Apologies Jorge,

The intention was to bring in the changes into master, on the rebuttable presumption that they've been reviewed already.
We don't have a lot of review bandwidth, so I wasn't expecting anyone to review the commits again, as we also wouldn't easily change them (maybe except the last commit with a `git commit --amend`.

After creating the PR, I noticed that the other merge options have been disabled in the UI (we can only squash now), so the aim of being able to use the UI to merge this was defeated; and so I rebased and pushed from my end.

I've gone through your questions around arraydata equality, so I'll have a look at what you've done on #8541, and I can prioritise any follow-up work needed to cover what might still be outstanding.

**Comment by jorgecarleitao** [comment]
> > There was probably some mis-understanding during the call yesterday, but I was expecting this to be a pull request and approved by another committer _before being merged_. This merge IMO did not follow our review practice and did not gave time to other committers to go through it.
> 
> Apologies Jorge,
> 
> The intention was to bring in the changes into master, on the rebuttable presumption that they've been reviewed already.
> We don't have a lot of review bandwidth, so I wasn't expecting anyone to review the commits again, as we also wouldn't easily change them (maybe except the last commit with a `git commit --amend`.
> 
> After creating the PR, I noticed that the other merge options have been disabled in the UI (we can only squash now), so the aim of being able to use the UI to merge this was defeated; and so I rebased and pushed from my end.
> 
> I've gone through your questions around arraydata equality, so I'll have a look at what you've done on #8541, and I can prioritise any follow-up work needed to cover what might still be outstanding.

No worries, was a mis-communication, and I am sorry for not being very explicit about it: I was aware of the writer branch, but since it was moving fast, I deemed it as "WIP" and waited for a PR to master before going through it. Once I saw the PR from @carols10cents to master, I understood it to be the "ready to review" version of the branch, and reviewed it ASAP, given the clock ticking on parquet stuff and the call.

Only afterwards I realize that that PR was actually a different thing, and that there has been a merge of the main changes already.

I would consider beneficial to have a review on any push to master from anyone. If anything, it gives time to others to comment on the PR and get accustomed to the changes.

**Comment by nevi-me** [comment]
That's okay, I understand.
I suppose it's becoming less of an issue as we seem to be growing commiters and reviewers. We'll also be doing the rest of the parquet work on the main branch, so this won't happen again.

**Comment by jorgecarleitao** [comment]
Anyways, I do not want this to downplay any of the work that mostly you and @carols10cents have been taking: the changes look really great. Thanks a lot for taking this on. It is a monumental effort and the outcome is great. 💯  Overall, this brings the rust implementation to its full potential of read and write to parquet.

**Pull request merged**